ethpm_cli/scraper.py

```diff
@@ -0,0 +1,195 @@
+from collections import OrderedDict
+import itertools
+import json
+from logging import Logger
+from pathlib import Path
+from typing import Any, Dict, Iterable, List, Set, Tuple  # noqa: F401
+
+from eth_utils import to_dict, to_hex, to_int, to_list
+from eth_utils.toolz import assoc
+from ethpm.typing import URI, Address
+from ethpm.utils.backend import resolve_uri_contents
+from ethpm.utils.ipfs import extract_ipfs_path_from_uri, is_ipfs_uri
+from ethpm.utils.uri import is_supported_content_addressed_uri
+from web3 import Web3
+from web3._utils.events import get_event_data
+
+from ethpm_cli._utils.various import flatten
+from ethpm_cli.constants import (
+    CONTENT_DIR,
+    EVENT_ABI as VERSION_RELEASE_EVENT_ABI,
+    TOPIC,
+)
+from ethpm_cli.exceptions import BlockNotFoundError, InstallError
+
+
+class Scraper:
+    """
+    Scrape blockchain for all valid 'VersionRelease' events
+    Resolves manifest uris & nested uris to a json object
+    """
+
+    def __init__(
+        self,
+        w3: Web3,
+        ws_w3: Web3 = None,
+        content_dir: Path = None,
+        logger: Logger = None,
+    ) -> None:
+        self.w3 = w3
+        # websocket w3 instance needed since infura requires ws for certain jsonrpc calls
+        self.ws_w3 = ws_w3 or w3
+        self.version_release_data: OrderedDict[str, Any] = OrderedDict()
+        self.last_processed_block = 0
```

Comment:
This statefulness gets bugs me a little, but I think that's mostly because I can see how we could do some small changes to remove it, but that may mean giving up on the scraper class, or at least extracting the functionality out of it.  Here's the gist.

```
class Config(NamedTuple):
    w3: Web3
    from_block: identifier  # maybe `None` signifies genesis
    to_block: identifier  # maybe `None` signifies to current


def process_uri(config, ipfs_uri) -> Set[Tuple[URI, Manifest]]:
    # download -> validate -> recurse-dependencies
    ...


def scrape_manifests_from_chain(config):
    # Low level function that is somewhat naive of statefulness of system.
    filter = w3.eth.filter(...)

    for log in filter:
        uri = get_uri_from_log(log)
        yield log, process_uri(config, uri)


def scrape(config, to_path: Path):
    # High level function that manages statefullness of system
    try:
        for log, manifests in scrape_chain(self.config):
            ...  
            # maybe write them to disk?
    except ...:
        # maybe save our progress (last processed log...)
```